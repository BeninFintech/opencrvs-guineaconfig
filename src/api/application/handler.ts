/*
 * Adaptation OpenCRVS pour la Guin√©e
 * Licence : MPL 2.0
 */
import { Request, ResponseToolkit } from '@hapi/hapi'
import { applicationConfig } from './application-config'
import { readCSVToJSON } from '@countryconfig/utils'

export interface Employee {
  primaryOfficeId: string
  givenNames: string
  familyName: string
  role: string
  mobile: string
  username: string
  email: string
  password: string
}

interface HealthFacility {
  id: string
  name: string
  partOf: string
  locationType: string
}

interface Location {
  id: string
  name: string
  jurisdictionType: string
  statisticalId: string
  partOf?: string 
}

interface Statistic {
  locationId: string
  femalePopulation: number
  malePopulation: number
  totalPopulation: number
  crude_birth_rate: number
}

// Helper pour garantir un tableau
function ensureArray<T>(data: T | T[]): T[] {
  return Array.isArray(data) ? data : [data]
}

// Handler pour les CRVS facilities
export async function crvsHandler(_: Request, h: ResponseToolkit) {
  const filePath = './src/data-seeding/locations/source/crvs-facilities.csv'
  const facilities = ensureArray(await readCSVToJSON<HealthFacility>(filePath))
  return h.response(facilities)
}

// Handler pour les health facilities 
export async function healthFacilitiesHandler(_: Request, h: ResponseToolkit) {
  const filePath = './src/data-seeding/locations/source/health_facilities.csv'
  const facilities = ensureArray(await readCSVToJSON<HealthFacility>(filePath))
  return h.response(facilities)
}

// Handler pour les locations
export async function locationsHandler(_: Request, h: ResponseToolkit) {
  const filePath = './src/data-seeding/locations/source/locations.csv'
  const locations = ensureArray(await readCSVToJSON<Location>(filePath))
  return h.response(locations) 
}

// Handler pour les statistics
export async function statisticsHandler(_: Request, h: ResponseToolkit) {
  const filePath = './src/data-seeding/locations/source/statistics.csv'
  const statistics = ensureArray(await readCSVToJSON<Statistic>(filePath))
  return h.response(statistics)
}

// Handler pour la configuration de l'application
export async function applicationConfigHandler(_: Request, h: ResponseToolkit) {
  return h.response(applicationConfig)
}
