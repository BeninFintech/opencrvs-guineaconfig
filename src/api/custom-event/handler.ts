/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * OpenCRVS is also distributed under the terms of the Civil Registration
 * & Healthcare Disclaimer located at http://opencrvs.org/license.
 *
 * Copyright (C) The OpenCRVS Authors located at https://github.com/opencrvs/opencrvs-core/blob/master/AUTHORS.
 */
import { env } from '@countryconfig/environment'
import { tennisClubMembershipEvent } from '@countryconfig/form/tennis-club-membership'
import { birthEvent } from '@countryconfig/form/v2/birth'
import { logger } from '@countryconfig/logger'
import * as Hapi from '@hapi/hapi'
import { createClient } from '@opencrvs/toolkit/api'
import { EventDocument } from '@opencrvs/toolkit/events'
import uuid from 'uuid'
import { customAlphabet } from 'nanoid'

const nanoid = customAlphabet('0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ', 12)

function generateRegistrationNumber() {
  return nanoid()
}

export function getCustomEventsHandler(
  _: Hapi.Request,
  h: Hapi.ResponseToolkit
) {
  return h.response([tennisClubMembershipEvent, birthEvent]).code(200)
}

export function onAnyActionHandler(
  request: Hapi.Request,
  h: Hapi.ResponseToolkit
) {
  // This catch-all event route can receive either legacy FHIR events with `Content-Type: application/fhir+json` or new events with `Content-Type: application/json`
  console.log(request.params.event, request.params.action)
  console.log(request.payload)
  console.log('RES CIHAN2', generateRegistrationNumber())
  console.log('RES CIHAN2', generateRegistrationNumber())
  console.log('RES CIHAN2', generateRegistrationNumber())
  return h.response().code(200)
}

export async function onRegisterHandler(
  _: Hapi.Request,
  h: Hapi.ResponseToolkit
) {
  // const event = EventDocument.parse(request.payload)
  // const client = createClient(
  //   env.GATEWAY_URL + '/events',
  //   request.headers.authorization
  // )

  // logger.info(`Confirming registration ${event.id}`)
  // await client.event.actions.register.accept.mutate({
  //   transactionId: uuid.v4(),
  //   eventId: event.id,
  //   data: {
  //     status: 'CONFIRMED'
  //   }
  // })
  // logger.info(`Registration ${event.id} confirmed`)

  return h
    .response({ registrationNumber: generateRegistrationNumber() })
    .code(200)
}
