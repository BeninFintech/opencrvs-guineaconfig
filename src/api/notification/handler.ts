/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * OpenCRVS is also distributed under the terms of the Civil Registration
 * & Healthcare Disclaimer located at http://opencrvs.org/license.
 *
 * Copyright (C) The OpenCRVS Authors located at https://github.com/opencrvs/opencrvs-core/blob/master/AUTHORS.
 */
import { logger, maskEmail, maskSms } from '@countryconfig/logger'
import * as Hapi from '@hapi/hapi'
import * as Joi from 'joi'
import { getApplicationConfig } from '../../utils'
import {
  COUNTRY_LOGO_URL,
  SENDER_EMAIL_ADDRESS,
  USER_NOTIFICATION_DELIVERY_METHOD
} from './constant'
import { sendEmail } from './email-service'
import { TriggerToSMSTemplate, sendSMS } from './sms-service'
import { TriggerVariable, getTemplate, renderTemplate } from './email-templates'

import {
  TriggerEvent,
  TriggerPayload,
  FullName
} from '@opencrvs/toolkit/notification'
import { LOGIN_URL } from '@countryconfig/constants'

type EmailPayloads = {
  subject: string
  html: string
  from: string
  to: string
}

export const emailSchema = Joi.object({
  subject: Joi.string().required(),
  html: Joi.string().required(),
  from: Joi.string().required(),
  to: Joi.string().required()
})

export async function emailHandler(
  request: Hapi.Request,
  h: Hapi.ResponseToolkit
) {
  const payload = request.payload as EmailPayloads

  if (process.env.NODE_ENV !== 'production') {
    logger.info(
      `Ignoring email due to NODE_ENV not being 'production'. Params: ${JSON.stringify(
        { ...payload, from: maskEmail(payload.from), to: maskEmail(payload.to) }
      )}`
    )
    return h.response().code(200)
  }

  await sendEmail(payload)

  return h.response().code(200)
}

export function makeNotificationHandler<T extends TriggerEvent>(event: T) {
  return async (request: Hapi.Request, h: Hapi.ResponseToolkit) => {
    const { payload: maybePayload } = request

    const payloadValidation = TriggerPayload[event].safeParse(maybePayload)

    if (!payloadValidation.success) {
      return h
        .response({
          error: 'Invalid payload',
          details: payloadValidation.error
        })
        .code(400)
    }

    await sendNotification(event, payloadValidation.data as TriggerPayload[T])

    return h.response().code(200)
  }
}

function stringifyV1Name(name: FullName) {
  return name.given[0] + ' ' + name.family
}

function generateFailureLog({
  contact,
  name,
  event,
  reason
}: {
  contact: { mobile?: string; email?: string }
  name: string
  event: TriggerEvent
  reason: string
}) {
  contact.mobile &&= maskSms(contact.mobile)
  contact.email &&= maskEmail(contact.email)
  logger.info(
    `Ignoring notification due to ${reason}. Params: ${JSON.stringify({
      event,
      recipient: {
        ...contact,
        name
      }
    })}`
  )
  return
}

export async function sendNotification<T extends TriggerEvent>(
  event: T,
  payload: TriggerPayload[T]
) {
  const variables = convertPayloadToVariable({
    event,
    payload
  } as TriggerEventPayloadPair)

  const { name: nameV1, ...contact } = payload.recipient
  const name = stringifyV1Name(nameV1[0])

  if (process.env.USER_NOTIFICATION !== 'true') {
    generateFailureLog({
      contact,
      name,
      event,
      reason: "USER_NOTIFICATION not being 'true'"
    })
    return
  }
  const applicationConfig = await getApplicationConfig()
  const applicationName = applicationConfig.APPLICATION_NAME

  if (USER_NOTIFICATION_DELIVERY_METHOD === 'email') {
    if (!payload.recipient.email) {
      generateFailureLog({
        contact,
        name,
        event,
        reason:
          "Not having recipient email when USER_NOTIFICATION_DELIVERY_METHOD is 'email'"
      })
      return
    }

    const template = getTemplate(event)

    const emailBody = renderTemplate(template, {
      ...variables,
      applicationName,
      countryLogo: COUNTRY_LOGO_URL
    })

    await sendEmail({
      subject: template.subject,
      html: emailBody,
      from: SENDER_EMAIL_ADDRESS,
      to: payload.recipient.email
    })
  } else if (USER_NOTIFICATION_DELIVERY_METHOD === 'sms') {
    if (!contact.mobile) {
      generateFailureLog({
        contact,
        name,
        event,
        reason:
          "Not having recipient mobile when USER_NOTIFICATION_DELIVERY_METHOD is 'sms'"
      })
      return
    }

    await sendSMS(
      TriggerToSMSTemplate[event],
      { ...variables, applicationName, countryLogo: COUNTRY_LOGO_URL },
      contact.mobile,
      'en'
    )
  } else {
    generateFailureLog({
      contact,
      name,
      event,
      reason: `Invalid USER_NOTIFICATION_DELIVERY_METHOD. Options are 'emai' or 'sms'. Found ${USER_NOTIFICATION_DELIVERY_METHOD}`
    })
    return
  }
}

export type TriggerEventPayloadPair<T extends TriggerEvent = TriggerEvent> = {
  [K in T]: {
    event: K
    payload: TriggerPayload[K]
  }
}[T]

function convertPayloadToVariable({
  event,
  payload
}: TriggerEventPayloadPair): TriggerVariable[typeof event] {
  const firstNames = payload.recipient.name[0].given[0]
  switch (event) {
    case TriggerEvent.USER_CREATED:
      return {
        firstNames: payload.recipient.name[0].given[0],
        username: payload.username,
        password: payload.temporaryPassword,
        completeSetupUrl: LOGIN_URL,
        loginURL: LOGIN_URL
      }

    case TriggerEvent.USER_UPDATED:
      return {
        firstNames,
        username: payload.username
      }

    case TriggerEvent.USERNAME_REMINDER:
      return {
        firstNames,
        username: payload.username
      }

    case TriggerEvent.RESET_PASSWORD:
      return {
        firstNames,
        authCode: payload.code
      }

    case TriggerEvent.RESET_PASSWORD_BY_ADMIN:
      return {
        firstNames,
        password: payload.temporaryPassword
      }

    case TriggerEvent.TWO_FA:
      return {
        firstNames,
        authCode: payload.code
      }

    default:
      throw new Error(`Unknown event: ${event}`)
  }
}
