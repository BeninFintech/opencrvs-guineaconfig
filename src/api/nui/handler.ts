/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * OpenCRVS is also distributed under the terms of the Civil Registration
 * & Healthcare Disclaimer located at http://opencrvs.org/license.
 *
 * Copyright (C) The OpenCRVS Authors located at https://github.com/opencrvs/opencrvs-core/blob/master/AUTHORS.
 */

import { Request } from '@hapi/hapi'
import { z } from 'zod'
import * as Joi from 'joi'
import {
  NUI_GENERATOR_URL,
  NUI_API_KEY,
  X_ROAD_CLIENT,
  CLIENT_SECRET,
  CLIENT_ID
} from './constants'
import { internal } from '@hapi/boom'

if (!NUI_GENERATOR_URL || !NUI_API_KEY) {
  throw new Error('NUI_GENERATOR_URL and NUI_API_KEY must be set')
}
const NUI_BATCH_SIZE = 1

export const nuiRequestBodySchema = Joi.object({
  office: Joi.string()
})

const generateNUIResponseSchema = z.object({
  theNuis: z.array(z.string()).min(NUI_BATCH_SIZE).max(NUI_BATCH_SIZE),
  status: z.string()
})

const affectNUIResponseSchema = z.object({
  status_code: z.number(),
  status: z.string()
})

const fetchNewToken = async (): Promise<string> => {
  try {
    const response = await fetch(`${NUI_GENERATOR_URL}/systems/token`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'X-Road-Client': X_ROAD_CLIENT
      },
      body: JSON.stringify({
        clientId: CLIENT_ID,
        clientSecret: CLIENT_SECRET
      })
    })

    const body: any = await response.json()

    if (!response.ok) {
      throw new Error(
        `Failed to fetch token: ${body.message || 'Unknown error'}`
      )
    }

    return body.data.accessToken
  } catch (err) {
    console.error('Error fetching new token', err)
    throw internal()
  }
}

export const generateNUI = async ({
  office,
  token
}: {
  office: string
  token?: string
}): Promise<string> => {
  try {
    const response = await fetch(`${NUI_GENERATOR_URL}/v1/generate`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer ${token}`,
        'X-Road-Client': X_ROAD_CLIENT
      },
      body: JSON.stringify({
        batchSize: NUI_BATCH_SIZE,
        destinator: office
      })
    })

    const body: any = await response.json()

    if (body.statusCode === 401) {
      token = await fetchNewToken()
      return await generateNUI({ office, token })
    }

    if (!response.ok) {
      throw new Error(
        `Failed to generate NUI: ${body.message || 'Unknown error'}`
      )
    }
    const parsed = generateNUIResponseSchema.parse(body)

    return parsed.theNuis[0]
  } catch (err) {
    console.error('Error generating NUI', err)
    throw internal()
  }
}

const affectNUI = async (nui: string, token: string) => {
  const response = await fetch(`${NUI_GENERATOR_URL}/v1/affect`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${token}`,
      'X-Road-Client': X_ROAD_CLIENT
    },
    body: JSON.stringify({
      nui: [nui]
    })
  })

  const body = await response.json()
  const parsed = affectNUIResponseSchema.parse(body)

  if (![200, 201].includes(parsed.status_code)) {
    console.error(parsed)
    throw internal()
  }
}

export async function NUIHandler(req: Request) {
  const body = req.payload as { office: string }
  try {
    const currentToken = await fetchNewToken()

    const nui = await generateNUI(body)
    await affectNUI(nui, currentToken)

    return nui
  } catch (err) {
    err.output.payload.message = 'httpError'
    throw err
  }
}
