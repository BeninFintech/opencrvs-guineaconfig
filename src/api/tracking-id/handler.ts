import * as Hapi from '@hapi/hapi'
import { generateTrackingId } from './service'

export function trackingIDHandler(
  request: Hapi.Request,
  h: Hapi.ResponseToolkit
) {
  const trackingId = generateTrackingId(request.params.eventType)
  return h.response(trackingId)
}
