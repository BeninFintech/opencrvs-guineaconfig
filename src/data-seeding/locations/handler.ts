/* eslint-disable prettier/prettier */
/*
 * Adaptation OpenCRVS pour la Guin√©e
 * Licence : MPL 2.0
 */
import {
  readCSVToJSON,
  extractStatisticsMap,
  getStatistics,
  LocationStatistic
} from '@countryconfig/utils'
import { Request, ResponseToolkit } from '@hapi/hapi'

type HumdataLocation = {
  admin0Pcode: string
  admin0Name_en: string
  admin0Name_alias?: string

  admin1Pcode?: string
  admin1Name_en?: string
  admin1Name_alias?: string

  admin2Pcode?: string
  admin2Name_en?: string
  admin2Name_alias?: string

  admin3Pcode?: string
  admin3Name_en?: string
  admin3Name_alias?: string
}

type Facility = {
  id: string
  name: string
  partOf: string
  locationType: 'HEALTH_FACILITY' | 'CRVS_OFFICE'
}

type Location = {
  id: string
  name: string
  alias: string
  partOf: string
  locationType: 'ADMIN_STRUCTURE' | 'HEALTH_FACILITY' | 'CRVS_OFFICE'
  jurisdictionType?: (typeof JURISDICTION_TYPE)[number]
  statistics?: LocationStatistic['years']
}

type ApplicationSetting = {
  General: string
  Configuration: string
}

const JURISDICTION_TYPE = [
  'COUNTRY',
  'REGION',
  'PREFECTURE',
  'SOUS_PREFECTURE'
] as const

// Helper pour forcer un retour tableau
function ensureArray<T>(data: T | T[]): T[] {
  return Array.isArray(data) ? data : [data]
}

export async function locationsHandler(_: Request, h: ResponseToolkit) {
  const [
    humdataLocations,
    healthFacilities,
    crvsFacilities,
    statistics
    // applicationSettings
  ] = await Promise.all([
    readCSVToJSON<HumdataLocation>(
      './src/data-seeding/locations/source/locations.csv'
    ).then(ensureArray),
    readCSVToJSON<Facility>(
      './src/data-seeding/locations/source/health_facilities.csv'
    ).then(ensureArray),
    readCSVToJSON<Facility>(
      './src/data-seeding/locations/source/crvs-facilities.csv'
    ).then(ensureArray),
    getStatistics(),
    // readCSVToJSON<ApplicationSetting>(
    //   './src/data-seeding/locations/source/application_setting.csv'
    // ).then(ensureArray)
  ])

  const locations = new Map<string, Location>()
  const statisticsMap = extractStatisticsMap(statistics)

  humdataLocations.forEach((loc) => {
    ;([0, 1, 2, 3] as const).forEach((level) => {
      const id = loc[`admin${level}Pcode` as keyof HumdataLocation] as string
      if (id) {
        const name = loc[`admin${level}Name_en` as keyof HumdataLocation] as string
        const alias =
          (loc[`admin${level}Name_alias` as keyof HumdataLocation] as string) ??
          name
        const parent =
          level === 0
            ? 'Location/0'
            : `Location/${loc[`admin${(level - 1) as 0 | 1 | 2}Pcode`]}`

        locations.set(id, {
          id,
          name,
          alias,
          partOf: parent,
          locationType: 'ADMIN_STRUCTURE',
          jurisdictionType: JURISDICTION_TYPE[level],
          statistics: statisticsMap.get(id)?.years
        })
      }
    })
  })

  ;[...healthFacilities, ...crvsFacilities].forEach((facility) => {
    locations.set(facility.id, {
      ...facility,
      alias: facility.name
    })
  })
  

  return h.response({
    locations: Array.from(locations.values()),
    applicationSettings
  })
}
