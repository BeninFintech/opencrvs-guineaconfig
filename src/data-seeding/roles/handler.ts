/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * OpenCRVS is also distributed under the terms of the Civil Registration
 * & Healthcare Disclaimer located at http://opencrvs.org/license.
 *
 * Copyright (C) The OpenCRVS Authors located at https://github.com/opencrvs/opencrvs-core/blob/master/AUTHORS.
 */
import { PRODUCTION, QA_ENV } from '@countryconfig/constants'
import roles from './roles.json'
import { Request, ResponseToolkit } from '@hapi/hapi'
import { readFile } from 'fs/promises'
import { join } from 'path'

export async function rolesHandler(_: Request, h: ResponseToolkit) {
  if (!PRODUCTION || QA_ENV) {
    const modifiableRoles: typeof roles = JSON.parse(
      (await readFile(join(__dirname, 'roles.json'))).toString()
    )

    return modifiableRoles.map((role) => {
      return {
        ...role,
        scopes: role.scopes.concat('demo')
      }
    })
  }
  return h.response(roles)
}
