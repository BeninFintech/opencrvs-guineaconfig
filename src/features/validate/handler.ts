/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * OpenCRVS is also distributed under the terms of the Civil Registration
 * & Healthcare Disclaimer located at http://opencrvs.org/license.
 *
 * Copyright (C) The OpenCRVS Authors. OpenCRVS and the OpenCRVS
 * graphic logo are (registered/a) trademark(s) of Plan International.
 */
import * as Hapi from '@hapi/hapi'
import fetch from 'node-fetch'
import * as Pino from 'pino'
import { CONFIRM_REGISTRATION_URL } from '@countryconfig/constants'
import { createWebHookResponseFromBundle } from '@countryconfig/features/validate/service'
import { badImplementation } from '@hapi/boom'

const logger = Pino()

export async function validateRegistrationHandler(
  request: Hapi.Request,
  h: Hapi.ResponseToolkit
) {
  try {
    const bundle = request.payload as fhir.Bundle

    const webHookResponse = await createWebHookResponseFromBundle(bundle)

    // This fetch can be moved to a custom task when validating externally
    fetch(CONFIRM_REGISTRATION_URL, {
      method: 'POST',
      body: JSON.stringify(webHookResponse),
      headers: request.headers
    })
  } catch (err) {
    /*

    I think that there might be a reason why we still call this endpoint even on an error
    But I think it is strange so I commented this out.

    I dont know what effect it is having yet

    fetch(CONFIRM_REGISTRATION_URL, {
      method: 'POST',
      body: JSON.stringify({ error: err.message }),
      headers: request.headers
    })*/

    logger.error(err)
    throw badImplementation(
      `Could not generate registration number in country configuration due to error: ${err}`
    )
  }

  return h.response().code(202)
}
