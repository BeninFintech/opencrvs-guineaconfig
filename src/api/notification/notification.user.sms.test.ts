import { vi, describe, it, expect, beforeEach } from 'vitest'
process.env.NODE_ENV = 'production'

vi.mock('../../utils', async (importOriginal) => {
  const actual = await importOriginal<typeof import('../../utils')>()
  return {
    ...actual,
    getApplicationConfig: vi.fn().mockResolvedValue({
      APPLICATION_NAME: 'Farajaland CRS',
      COUNTRY: 'BD',
      COUNTRY_LOGO: { url: '/logo.png' },
      SENTRY: 'https://sentry.com',
      LOGIN_BACKGROUND: { url: '/bg.png' },
      USER_NOTIFICATION_DELIVERY_METHOD: 'sms',
      INFORMANT_NOTIFICATION_DELIVERY_METHOD: 'sms'
    })
  }
})

vi.mock(import('./constant'), async (importOriginal) => {
  const actual = await importOriginal()
  return {
    ...actual,
    INFOBIP_API_KEY: 'mock_api_key',
    INFOBIP_GATEWAY_ENDPOINT: 'https://gateway.infobip.com',
    INFOBIP_SENDER_ID: 'mock_sender'
  }
})

vi.mock('node-fetch', () => {
  return {
    default: vi.fn().mockResolvedValue({
      ok: true,
      status: 200,
      json: async () => ({})
    })
  }
})

import { sendNotification } from './handler'
import fetch from 'node-fetch'

describe('User notification - sms', () => {
  it('calls fetch with correct params', async () => {
    await sendNotification('user-created', {
      recipient: {
        name: [{ use: 'en', family: 'Doe', given: ['John'] }],
        email: 'john.doe@gmail.com',
        mobile: '+15551234567'
      },
      username: 'j.doe',
      temporaryPassword: 'TempPass123!'
    }).catch(() => {})

    expect((fetch as any).mock.calls[0][1].body).toMatchSnapshot()
  })
})
