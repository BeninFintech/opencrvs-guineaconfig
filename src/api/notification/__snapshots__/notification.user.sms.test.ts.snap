// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`User notification - sms > calls fetch with correct params 1`] = `"{"messages":[{"destinations":[{"recipient":"+15551234567"}],"from":"mock_sender","text":"Your OpenCRVS account has been created. Please login with your username: j.doe and temporary password: TempPass123!"}]}"`;
